aries_vcx/src/new_protocols/connection/invitee/handlers.rs

```diff
@@ -0,0 +1,217 @@
+            let service = match get_service(ledger, &invitation.content.service_endpoint).await {
+                Ok(s) => s,
+                Err(err) => {
+                    error!("Failed to obtain service definition from the ledger: {}", err);
+                    return Err(err);
+                }
+            };
+
+            // See https://github.com/hyperledger/aries-rfcs/blob/main/features/0160-connection-protocol/README.md#agency-endpoint
+            invitation.content.routing_keys.extend(service.recipient_keys);
+
+            (
+                service.service_endpoint,
+                invitation.content.recipient_keys,
+                invitation.content.routing_keys,
+                None,
+                Some(invitation.content.service_endpoint),
+            )
+        }
+    };
+
+    let bootstrap_info = BootstrapInfo {
+        service_endpoint,
+        recipient_keys,
+        routing_keys,
+        did,
+        service_endpoint_did,
+    };
+
+    Ok(bootstrap_info)
+}
+
+// TODO: This won't accept this many args in its final version
+// and should process the invitation in a better way
+pub async fn accept_invitation<S, W>(
+    sm_storage: S,
+    sm_id: S::Id,
+    invitation: Invitation,
+    service_endpoint: Url,
+    routing_keys: Vec<String>,
+    label: String,
+    ledger: &Arc<dyn IndyLedgerRead>,
+    wallet: &W,
+) -> VcxResult<Request>
+where
+    S: StateMachineStorage,
+    W: BaseWallet,
+{
+    let msg_id = Uuid::new_v4().to_string();
+
+    let thread = match &invitation {
+        Invitation::Public(i) => {
+            let mut thread = Thread::new(msg_id.clone());
+            thread.pthid = Some(i.id.clone());
+            thread
+        }
+        Invitation::Pairwise(i) => Thread::new(i.id.clone()),
+        Invitation::PairwiseDID(i) => Thread::new(i.id.clone()),
+    };
+
+    let bootstrap_info = did_doc_from_invitation(ledger, invitation).await?;
+    let (did, verkey) = wallet.create_and_store_my_did(None, None).await?;
+
+    let recipient_keys = vec![verkey.clone()];
+
+    let mut did_doc = AriesDidDoc::default();
+    did_doc.id = did.clone();
+    did_doc.set_service_endpoint(service_endpoint);
+    did_doc.set_routing_keys(routing_keys);
+    did_doc.set_recipient_keys(recipient_keys);
+
+    let con_data = ConnectionData::new(did.clone(), did_doc);
+
+    let (sm, content) =
+        InviteeConnection::new_invitee(did, verkey, label, bootstrap_info, con_data, thread.thid.clone());
+
+    let timing = Timing {
+        out_time: Some(Utc::now()),
+        ..Default::default()
+    };
+
+    let decorators = RequestDecorators {
+        thread: Some(thread),
+        timing: Some(timing),
+    };
+
+    let request = Request::with_decorators(msg_id, content, decorators);
+    let sm = AriesSM::Connection(ConnectionSM::InviteeRequested(sm));
+    sm_storage.put_new_state(sm_id, sm).await?;
+
+    Ok(request)
+}
+
+pub async fn handle_response<S, W>(
```

Comment:
### Opinion:

It tries to shield away user from concept of states and state machines, and worry about the result because the data just gets update in the storage. But how will user handle following cases then: 
- the transitioning function fail with error which might disappear upon retry - typically any IO - for example resolving DID Document. The FSM in the storage stay the same, an `Error` is returned from the function. Is every `Error` supposed to signal retryable operation?
- Is `Ok` variant of the function always the "happy case"? What would be approach for entering "hard failed" states? It seems like there 2 options:
  - If you return Err, we'll need to distinguish retryable errs and protocols erros (not retryable)
  - if you return Ok, the user won't know what actually happened in transition - whether the state machine update in the storage behind the scenes - in our transition function - ended up in "happy case" state or "non-happy case" state.

Reply:
### Opinion

Since these - essentially transition functions / processing functions - are divorced from state machines themselves, it makes state machines a bit more universal - technically someone could implement their own way of, for example - processing connection request.
In old paradigm the code existed directly on the state machines, and you either take it or not.